OpenRA.Renderer.Sdl2/Sdl2GraphicsDevice.cs

```diff
@@ -330,6 +330,42 @@ public void SetLineWidth(float width)
 			ErrorHandler.CheckGlError();
 		}
 
+		public Bitmap TakeScreenshot()
+		{
+			var rect = new Rectangle(Point.Empty, size);
+			var bitmap = new Bitmap(rect.Width, rect.Height, System.Drawing.Imaging.PixelFormat.Format32bppArgb);
+			var data = bitmap.LockBits(rect,
+				System.Drawing.Imaging.ImageLockMode.WriteOnly, System.Drawing.Imaging.PixelFormat.Format32bppArgb);
+
+			try
+			{
+				int packAlignment, unpackAlignment;
+
+				unsafe
+				{
+					GL.GetInteger(GetPName.PackAlignment, out packAlignment);
+					GL.GetInteger(GetPName.UnpackAlignment, out unpackAlignment);
+				}
+
+				GL.PixelStore(PixelStoreParameter.UnpackAlignment, 1);
+				GL.PixelStore(PixelStoreParameter.PackAlignment, 1);
+
+				GL.ReadPixels(rect.X, rect.Y, rect.Width, rect.Height, PixelFormat.Bgra, PixelType.UnsignedByte, data.Scan0);
+				GL.Finish();
+
+				GL.PixelStore(PixelStoreParameter.UnpackAlignment, unpackAlignment);
```

Comment:
You can just safely leave the alignment at 1.

On Mon, Jun 8, 2015 at 10:08 AM, Matija Hustić <notifications@github.com>
wrote:

> In OpenRA.Renderer.Sdl2/Sdl2GraphicsDevice.cs
> <https://github.com/OpenRA/OpenRA/pull/7930#discussion_r31881281>:
>
> > +			{
> > +				int packAlignment, unpackAlignment;
> > +
> > +				unsafe
> > +				{
> > +					GL.GetInteger(GetPName.PackAlignment, out packAlignment);
> > +					GL.GetInteger(GetPName.UnpackAlignment, out unpackAlignment);
> > +				}
> > +
> > +				GL.PixelStore(PixelStoreParameter.UnpackAlignment, 1);
> > +				GL.PixelStore(PixelStoreParameter.PackAlignment, 1);
> > +
> > +				GL.ReadPixels(rect.X, rect.Y, rect.Width, rect.Height, PixelFormat.Bgra, PixelType.UnsignedByte, data.Scan0);
> > +				GL.Finish();
> > +
> > +				GL.PixelStore(PixelStoreParameter.UnpackAlignment, unpackAlignment);
>
> Just trying to understand the code here - is there a reason why you get
> and set the UnpackAlignment?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/OpenRA/OpenRA/pull/7930/files#r31881281>.
>